Adding motion tracking example. (#40)
zabertech/zaber-examples: examples/motion_tracking/README.md

# Tracking Motion Using PID Controller

> This example demonstrates how Zaber stages can track analog input signals using a PID controller.

Reviewer: ```suggestion
This example demonstrates how Zaber stages can track analog input signals using a PID controller,
```

> PID controller is a control loop feedback mechanism widely used in industrial control systems.

Reviewer: ```suggestion
which is a control loop feedback mechanism widely used in industrial control systems.
```

Zaber only supports the PI controller, which is a simplified version of the PID controller without derivative term.

The following video demonstrates the concept:

![video.mp4](img/video.mp4)

In the video above there is a load cell mounted on a Zaber stage.
The stage is tracking the analog input signal from the load cell keeping a constant tension on the spring.
When disrupted, the stage moves to compensate for the change in the force until it reaches the setpoint again.

## Code

### Dependencies

The script uses `pdm` to manage the virtual environment and dependencies:

Instructions on how to install it can be found on the official `pdm` project page [here](https://github.com/pdm-project/pdm).

The dependencies are listed in `pyproject.toml`.

### Running the Script

Make sure to change `SERIAL_PORT` constant in `main.py` file to the serial port that your device is connected to.

Afterwards, you can run the example:

```shell
cd examples/motion_tracking/
pdm install
pdm run example
```

### Explanation

<https://github.com/zabertech/zaber-examples/blob/main/examples/motion_tracking/src/motion_tracking/main.py#L18-L33>

> After establishing a connection with the device, the script prints PI controller parameters.

Reviewer: ```suggestion
After establishing a connection with the device, the script prints the PI controller parameters.
```

Then based on the user input "track" the scripts starts the PI controller by issuing the `move track` command to the axis.
After that, the stage will track the analog input signal indefinitely until the user inputs "stop".

## Tuning the Controller

Properly tuning the controller requires a basic understanding of control theory and is beyond the scope of this example.

> Typically, you adjust the Kp and Ki parameters to achieve desired response.

Reviewer: ```suggestion
Typically, you adjust the Kp and Ki parameters to achieve the desired response.
```
